src/windows/network.rs

```diff
@@ -246,4 +287,137 @@ impl NetworkExt for NetworkData {
     fn total_errors_on_transmitted(&self) -> u64 {
         self.errors_out
     }
+
+    fn mac_address(&self) -> MacAddr {
+        self.mac_addr
+    }
+
+    fn ipv4_address(&self) -> Ipv4Addr {
+        self.ipv4_addr
+    }
+
+    fn ipv4_netmask(&self) -> Ipv4Addr {
+        self.ipv4_mask
+    }
+}
+
+pub(crate) struct InterfaceAddressIterator {
+    /// The first item in the linked list
+    buf: PIP_ADAPTER_ADDRESSES,
+    /// The current adapter
+    adapter: PIP_ADAPTER_ADDRESSES,
+    /// IP addresses grouped by current adapter
+    unicast_address: PIP_ADAPTER_UNICAST_ADDRESS,
+}
+
+// Need a function to convert u16 pointer into String
+// https://stackoverflow.com/a/48587463/8706476
+unsafe fn u16_ptr_to_string(ptr: *const u16) -> OsString {
+    let len = (0..).take_while(|&i| *ptr.offset(i) != 0).count();
+    let slice = std::slice::from_raw_parts(ptr, len);
+
+    OsString::from_wide(slice)
+}
+
+impl Iterator for InterfaceAddressIterator {
+    type Item = (String, InterfaceAddress);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.adapter.is_null() {
+            return None;
+        }
+        unsafe {
+            let adapter = self.adapter;
+            if let Ok(interface_name) = u16_ptr_to_string((*adapter).FriendlyName).into_string() {
+                if self.unicast_address.is_null() {
+                    // if we found that this interface has not been visited yet,
+                    // set unicast address,
+                    self.unicast_address = (*adapter).FirstUnicastAddress;
+                    // and return the MAC address instead
+                    let [mac @ .., _, _] = (*adapter).PhysicalAddress;
+                    Some((
+                        interface_name,
+                        InterfaceAddress::MAC(MacAddr::from(mac)),
+                    ))
+                } else {
+                    // otherwise, generate an IP adddress
+                    let address = self.unicast_address;
+                    self.unicast_address = (*address).Next;
+                    if self.unicast_address.is_null() {
+                        // if we have visited all unicast addresses, move to next adapter
+                        self.adapter = (*adapter).Next;
+                    }
+                    // FIXME: should we perform a null check on (*address).Address.lpSockaddr
```

Comment:
Yes, most definitely!

Reply:
It is placed in `src/network_helper_win.rs` now. If the pointer was null, call `self.next()`.